### PR TITLE
[Merged by Bors] - p2p/peers: ensure that peer events are received by peer manager

### LINKED ISSUE #2385: Drop peers after failing to send messages to them
## Description
I'm seeing ~200k/hr p2p warnings about "failed sending" messages from the managed testnet nodes (see screenshot below). This happens when a node tries to propagate a gossip message to a peer, and fails. I think this error is happening so often because nodes never drop peers after such an error. Ideally, a node should drop a peer--in the case of connection not found in cpool, probably right away, and replace it with a healthy peer.

## Affected code

https://github.com/spacemeshos/go-spacemesh/blob/a3ba021480a2aa5ed9b88d9d2c1f62dbddf7c1f1/p2p/gossip/protocol.go#L145

This issue appears in commit hash: v0.1.29, a3ba021480a2aa5ed9b88d9d2c1f62dbddf7c1f1

Related files (optionally with line numbers): 

image

### PR DESCRIPTION
**Pull request by dshulyak**
## Motivation
Sometimes connections are removed from cpool but peers are not removed from peers store that is formed by listening to notifications. Notifications were non-blocking, so in certain goroutine scheduling scenario it was possible that peers event listener will miss some events. I am not totally sure that this was the case, but I don't see any other problem in this code. 

closes: https://github.com/spacemeshos/go-spacemesh/issues/2385

## Changes
- make peer notifications blocking with a buffer that can fit a potential burst of connected or expired peers 
- don't copy and stop shuffling peers when GetPeers is called, this adds a significant pressure on gc for nothing
  there are 3 cases when GetPeers is used, in two of them all peers are receiving messages, and in the other one random peer is selected from the array

## Test Plan
TODO

## TODO
- [x] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed

## DevOps Notes
- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
@antonlerner take a look please, i will experiment a bit with this change on real environment before moving forward. but maybe you will have some other suggestions

File: p2p/switch.go [hunk 1]
```diff
@@ -677,7 +677,7 @@ func (s *Switch) publishNewPeer(peer p2pcrypto.PublicKey) {
 	for _, p := range s.newPeerSub {
 		select {
 		case p <- peer:
-		default:
+		case <-s.shutdownCtx.Done():
```
Comment (antonlerner) [inline p2p/switch.go:680]: return here?

**Comment by antonlerner** [comment]
Good catch of this edge case
Another thing is to remove the duplicate peer lists we have throughout the p2p code, but it is a lot to refactor so I suggest we test this fix and see if it reduces the p2p errors seen

**Comment by dshulyak** [comment]
> Another thing is to remove the duplicate peer lists we have throughout the p2p code, but it is a lot to refactor so I suggest we test this fix and see if it reduces the p2p errors seen

@antonlerner can you show me what you have in mind? i thought that all calls were proxied to Peers object

**Comment by antonlerner** [comment]
> > Another thing is to remove the duplicate peer lists we have throughout the p2p code, but it is a lot to refactor so I suggest we test this fix and see if it reduces the p2p errors seen
> 
> @antonlerner can you show me what you have in mind? i thought that all calls were proxied to Peers object

this is the duplication I'm talking about, there are inbound and outbound lists of peers in `switch.go`, IMO there is no real need to keep another list in `peers.go`

**Review by antonlerner — APPROVED** [review]

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors try

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
build merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge

**Comment by dshulyak** [comment]
bors merge
